### PR TITLE
Auth/PM-7811 - Refactor User Auto Unlock Key Hydration Process To Remove Race Conditions

### PR DESCRIPTION
**Pull request by JaredSnider-Bitwarden**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

To resolve [PM-7811 ](https://bitwarden.atlassian.net/browse/PM-7811) which initially presented itself as the CLI not properly unlocking after exporting a session key on a fresh install or after deleting `data.json`. During our investigation, we realized there was a cross client race condition with any code that called to check if the user had a user key in memory and the `UserKeyInitService's` listener for setting the user key in memory if the user auto unlock key was set. This refactor should eliminate any chances of race conditions and ensure that the user key in memory is always set as early as possible in the client initialization processes.

## Code changes

- Renamed `UserKeyInitService` to `UserAutoUnlockKeyService` as it is now purely involved with managing the user auto unlock key. 
- Updated `UserAutoUnlockKeyService` dependencies
- **apps/web/src/app/core/init.service.ts:** Added `userAutoUnlockKeyService.setUserKeyInMemoryIfAutoUserKeySet` on Web client init for singular active account
- **apps/desktop/src/app/services/init.service.ts:** Added `userAutoUnlockKeyService.setUserKeyInMemoryIfAutoUserKeySet` on Desktop client init for n accounts (desktop supports account switching)
- **apps/browser/src/background/main.background.ts:** Added `userAutoUnlockKeyService.setUserKeyInMemoryIfAutoUserKeySet` on Browser Extension client init for n accounts (browser extension supports account switching)
- **apps/cli/src/bw.ts:** Added `userAutoUnlockKeyService.setUserKeyInMemoryIfAutoUserKeySet` on CLI client init for singular active account which executes pre-command execution. 

### Web

https://github.com/bitwarden/clients/assets/116684653/a8abcf41-a0c6-4b67-abec-c394ec5389dd

### CLI

PM-7811 - CLI - never lock works

### Desktop

https://github.com/bitwarden/clients/assets/116684653/c062a02e-8e79-400c-aca1-f13d1befd9f9

### Browser Extension

https://github.com/bitwarden/clients/assets/116684653/0423c810-ac74-489d-932b-5c66db9d4048

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: apps/browser/src/auth/popup/account-switching/services/account-switcher.service.ts [hunk 1]
```diff
@@ -160,5 +162,8 @@ export class AccountSwitcherService {
       }
       throw err;
     });
+
+    // Set the user key in memory if the auto user key is set on acct switch
+    await this.userAutoUnlockKeyService.setUserKeyInMemoryIfAutoUserKeySet(id as UserId);
```
Comment (justindbaur) [inline apps/browser/src/auth/popup/account-switching/services/account-switcher.service.ts:167]: I think with desktop we realized this doesn't actually need to be done here right? Signing into an account with never lock and it being in background should be enough.

**Review by justindbaur — COMMENTED** [review]
Just one question

File: apps/browser/src/auth/popup/account-switching/services/account-switcher.service.ts (hunk 1 above)
Comment (JaredSnider-Bitwarden) [reply apps/browser/src/auth/popup/account-switching/services/account-switcher.service.ts]: Oops, you're right. I've fixed it and retested to confirm it was working as expected.

**Review by justindbaur — APPROVED** [review]

File: apps/browser/src/auth/popup/account-switching/services/account-switcher.service.ts (hunk 1 above)
Comment (justindbaur) [reply apps/browser/src/auth/popup/account-switching/services/account-switcher.service.ts]: Thanks, looks good!

**Review by jlf0dev — APPROVED** [review]

**Pull request merged**